### DISCUSSION TITLE
support additional/custom properties

### DISCUSSION THREAD

**Discussion by Cyclonit** [likec4/likec4#1342] [Ideas]
# Problem

Currently LikeC4 supports three properties for nodes: `title`, `description`, and `technology` ([documentation](https://likec4.dev/dsl/model/#element-properties). This schema was adopted from Structurizr. Edges support the same properties.

![grafik](https://github.com/user-attachments/assets/c0386316-2a8c-4128-b436-b7bc599a2f8e)
![grafik](https://github.com/user-attachments/assets/90a43aa4-c658-4a73-8752-a70369b99203)

Title and description are self explanatory, but the choice of making "technology" the only additional is somewhat opinionated. I'd like to discuss whether we can open up this system, to support additional or perhaps entirely custom properties for both nodes and edges.

# Examples

- We would like to track encrypted vs non-encrypted edges. Thus being able to add a property `encrypted = true/false` to each node would be nice.
- Being able to track `push` vs `pull` connections.

**Comment by wafelmolt** [comment]
Actually, you can yet add `metadata` to your c4 elements as Key/Value elements. Maybe the right way would be to flag independently those MD so one can choose to display it on the diagram.

**Reply by kieronlanning** [reply]
Is it possible to have array/ collections on `metadata`?

I'm trying to model some 'warnings' or 'issues' on some services.

At the moment, I'm doing this:

```
#Warning

metadata {
    warning_1 'Here is some information about the first warning'
    warning_2 'Here is another warning'
}
```

What I'd _like_ to see is:

```
#Warning

metadata {
    warning 'Here is some information about the first warning'
    warning 'Here is another warning'
}
```

This would get turned into an array/ collection for display. The advantage of this over forcing an array syntax (`['some text', ...]`) is as you add or remove and there's a single element, you wouldn't have to turn it into a normal key/ value pair.

// @davydkov

**Reply by kieronlanning** [reply]
Begun using this feature more and more over the past few days, and actually having a nicer interface (table?) means of interaction would be a nice addition @davydkov

**Reply by davydkov** [reply]
@kieronlanning v1.33 brings some UI improvements.

<img alt="image" src="https://github.com/user-attachments/assets/320f89c3-43bd-4a27-b744-9947898452ff" />
 
But adding support for arrays in metadata values will take a bit more work.

**Reply by kieronlanning** [reply]
Amazing, Thankyou!
